pkg/migrator/transformer.go

```diff
@@ -0,0 +1,341 @@
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/xmtp/xmtpd/pkg/constants"
+	"github.com/xmtp/xmtpd/pkg/envelopes"
+	"github.com/xmtp/xmtpd/pkg/proto/identity/associations"
+	mlsv1 "github.com/xmtp/xmtpd/pkg/proto/mls/api/v1"
+	proto "github.com/xmtp/xmtpd/pkg/proto/xmtpv4/envelopes"
+	"github.com/xmtp/xmtpd/pkg/topic"
+	"github.com/xmtp/xmtpd/pkg/utils"
+	protobuf "google.golang.org/protobuf/proto"
+)
+
+// TODO: Validate payer envelope.
+// TODO: Validate key package.
+
+type transformer struct {
+	payerPrivateKey *ecdsa.PrivateKey
+	nodeSigningKey  *ecdsa.PrivateKey
+}
+
+func NewTransformer(
+	payerPrivateKey *ecdsa.PrivateKey,
+	nodeSigningKey *ecdsa.PrivateKey,
+) *transformer {
+	return &transformer{
+		payerPrivateKey: payerPrivateKey,
+		nodeSigningKey:  nodeSigningKey,
+	}
+}
+
+func (t *transformer) Transform(record ISourceRecord) (*envelopes.OriginatorEnvelope, error) {
+	switch record.TableName() {
+	case groupMessagesTableName:
+		data, ok := record.(*GroupMessage)
+		if !ok {
+			return nil, fmt.Errorf("invalid record type: %T", record)
+		}
+
+		return t.TransformGroupMessage(data)
+
+	case inboxLogTableName:
+		data, ok := record.(*InboxLog)
+		if !ok {
+			return nil, fmt.Errorf("invalid record type: %T", record)
+		}
+
+		return t.TransformInboxLog(data)
+
+	case installationsTableName:
+		data, ok := record.(*Installation)
+		if !ok {
+			return nil, fmt.Errorf("invalid record type: %T", record)
+		}
+
+		return t.TransformInstallation(data)
+
+	case welcomeMessagesTableName:
+		data, ok := record.(*WelcomeMessage)
+		if !ok {
+			return nil, fmt.Errorf("invalid record type: %T", record)
+		}
+
+		return t.TransformWelcomeMessage(data)
+
+	default:
+		return nil, fmt.Errorf(
+			"Transform not implemented for table: %s",
+			record.TableName(),
+		)
+	}
+}
+
+// TransformGroupMessage converts GroupMessage to appropriate XMTPD envelope format.
+func (t *transformer) TransformGroupMessage(
+	groupMessage *GroupMessage,
+) (*envelopes.OriginatorEnvelope, error) {
+	if groupMessage == nil {
+		return nil, fmt.Errorf("groupMessage is nil")
+	}
+
+	if groupMessage.GroupID == nil {
+		return nil, fmt.Errorf("groupID is nil")
+	}
+
+	if len(groupMessage.Data) <= 0 {
+		return nil, fmt.Errorf("data is empty")
+	}
+
+	protoClientEnvelope := &proto.ClientEnvelope{
+		Payload: &proto.ClientEnvelope_GroupMessage{
+			GroupMessage: &mlsv1.GroupMessageInput{
+				Version: &mlsv1.GroupMessageInput_V1_{
+					V1: &mlsv1.GroupMessageInput_V1{
+						Data: groupMessage.Data,
+					},
+				},
+			},
+		},
+		Aad: &proto.AuthenticatedData{
+			TargetTopic: topic.NewTopic(topic.TOPIC_KIND_GROUP_MESSAGES_V1, groupMessage.GroupID[:]).
+				Bytes(),
+		},
+	}
+
+	return t.originatorEnvelope(
+		protoClientEnvelope,
+		groupMessageOriginatorID,
+		uint64(groupMessage.ID),
+	)
+}
+
+// TransformInboxLog converts InboxLog to appropriate XMTPD IdentityUpdate envelope format.
+func (t *transformer) TransformInboxLog(
```

Comment:
When we process an identity update from the blockchain (in the old node and new) we update the address log as well when addresses are added and removed from an inbox as part of an update. 

We will need to do that somewhere in this process. We could pull from the address_log table separately, but it might become difficult to reconcile with identity updates coming in from the chain. Especially since rows in the address log are mutable.

The alternative is to do the same thing we do when pulling from the chain, and get the list of added/revoked addresses by processing the update through the MLS validation service. Adds an external service call, but is closer to how this process works everywhere else.

Reply:
My idea is to inject the MLS validation service in the writer and doing it there. 
Same as in the indexer; when processing an IdentityUpdate, get the association states and act based on the diffs.

The difference is that in the indexer we get the gateway envelopes with originatorID 1 (IU from chain), and in the migrator writer we'd get the originatorID 12 since AFAIK these identity updates would only be relevant to the ones being migrated.